robot/web/web.go

```diff
@@ -365,42 +364,44 @@ func (svc *webService) updateResources(resources map[resource.Name]resource.Reso
 		groupedResources[n.API] = r
 	}
 
-	apiRegs := resource.RegisteredAPIs()
-	for a, v := range groupedResources {
-		apiResColl, ok := svc.services[a]
-		// TODO(RSDK-144): register new service if it doesn't currently exist
+	// All known APIs have pre-registered resource collections, so loop through them. We want to empty out any collections if
+	// there are no longer resources available in that API.
+	for api, coll := range svc.services {
+		group, ok := groupedResources[api]
 		if !ok {
-			reg, ok := apiRegs[a]
-			var apiResColl resource.APIResourceCollection[resource.Resource]
-			if ok {
-				apiResColl = reg.MakeEmptyCollection()
-			} else {
-				// Log a warning here to remind users to register their APIs. Do not warn if the resource is internal to the RDK or
-				// the resource is handled by a remote with a possibly separate API registration. Modular resources will
-				// have API registrations already and should not reach this point in the method.
-				if a.Type.Namespace != resource.APINamespaceRDKInternal {
-					for n := range v {
-						if !n.ContainsRemoteNames() {
-							svc.logger.Warnw(
-								"missing registration for api, resources with this API will be unreachable through a client", "api", n.API)
-							break
-						}
-					}
+			// create an empty map of resources if one does not exist
+			group = make(map[resource.Name]resource.Resource)
+		}
+		if err := coll.ReplaceAll(group); err != nil {
+			return err
+		}
+		delete(groupedResources, api)
+	}
+
+	// If there are any groupedResources remaining, check if they are registered/internal/remote.
```

Comment:
This documentation is excellent.

I'm going to mention (but not ask for any changes -- just curious if there's a specific disagreement): I think this is a bit of an unnatural way to log/warn when we run into resources that are part of "unexpected" APIs.

The above loop walks registered service APIs and pulls in the resources from the input `groupedResources` that apply. And deletes the grouped resources, potentially leaving some dangling entries. Which the below loop iterates over.

I would expect that we instead loop through the `groupedResources` and as we walk each API -- if the web service knows of it, copy those resources over.

And if the web service does not* recognize the API, run this battery of checks of whether its worth logging (probably wrapped up into a function to keep the above, more important mechanism streamlined/remove distractions for the reader).

I did pick this up easily enough (again not asking for a change here). The most confusing thing was the delete on 378 being unjustified. And having to grok this comment to see why that was done. Not knowing this comment would be answering that question.

Reply:
the original loop did loop through `groupedResources` but then that causes the issue of not iterating through apis that do not have resources and failing to empty the collection out. I think the general flow also ends up being a bit confusing and obfuscated as well, which is why I changed it to more of a 2 phase operation

depending on what we loop over, we still have to keep track of something (with looping through `groupedResources` we have to check which APIs we haven't looped over, and in the current case we have to keep track of the dangling APIs that don't have collections for). Maybe I can leave a comment above the delete?

Reply:
Ah, with that in my mind and rereading the documentation on [line 367](https://github.com/viamrobotics/rdk/pull/4398/files#diff-c53b2d6b81f57688a657b56a6d9b30f7e072d22ae5b709e39b4e05c8838bd5c4R367), I totally missed the nuance. So the bug wasn't that a* resource was removed. But if all resources for a given API were removed, we'd miss that. I'm going to suggest a comment that I'd (maybe) have picked up on.